lib/base-compiler.js

```diff
@@ -640,201 +701,70 @@ class BaseCompiler {
-                                }
-                                return result;
-                            })
-                            .then(result => {
-                                result.popularArguments = this.possibleArguments.getPopularArguments(options);
-
-                                return result;
-                            })
-                            .then(result => {
-                                if (result.okToCache) {
-                                    this.env.cachePut(key, result);
-                                }
-
-                                return Promise.all([executionPromise]).then(([executionResult]) => {
-                                    if (executionResult) {
-                                        result.execResult = executionResult;
-                                    }
-
-                                    return result;
-                                });
-                            });
-                    } else {
-                        return this.handleExecution(key, executeParameters);
-                    }
+            // TODO make const when I can
+            let [result, optOutput] = await this.doCompilation(
+                inputFilename, dirPath, key, options, filters, backendOptions, libraries, tools);
+
+            if (result.hasOptOutput) {
+                delete result.optPath;
+                result.optOutput = optOutput;
+            }
+            result.tools = _.union(result.tools, await Promise.all(this.runToolsOfType(tools, "postcompilation",
+                this.getCompilationInfo(key, result))));
+            if (result.dirPath) {
+                fs.remove(result.dirPath);
+                result.dirPath = undefined;
+            }
+            if (result.okToCache) {
+                result.asm = this.processAsm(result, filters);
+            } else {
+                result.asm = [{text: result.asm}];
+            }
+            // TODO rephrase this so we don't need to reassign
+            result = filters.demangle ? await this.postProcessAsm(result, filters) : result;
+            if (this.compiler.supportsCfg && backendOptions && backendOptions.produceCfg) {
+                result.cfg = cfg.generateStructure(this.compiler.compilerType,
+                    this.compiler.version, result.asm);
+            }
+            result.popularArguments = this.possibleArguments.getPopularArguments(options);
+            if (result.okToCache) {
+                this.env.cachePut(key, result).then(() => {
+                    // purely here to ensure the cache put happens (if it's not "then"-ned it might not start. as it
+                    // happens it does cos internally to cachePut it "catch"-es. but I don't want to rely on this.
                 });
-            });
+            }
+
+            if (doExecute) {
+                result.execResult = await this.handleExecution(key, executeParameters);
```

Comment:
moving this down here is potentially a performance pessimization, since all the tools, demanglers, and post processing will have to run to completion before we even start executing.

I am more familiar with async/await patterns in languages like python and C#, but at least in those languages a not-yet-awaited task will still run to completion behind the scenes as soon as it is created.

Its likely a small difference if any in most cases, but worth thinking about.

Reply:
Good point. My assumption (which is probably wrong) is that a promise doesn't start running until you await() it. In the specific case of executing another external process that might not be the case...but it's worth looking in to. Thanks

Reply:
I'm still not 100% on this but I changed it back as it was a simple enough change.